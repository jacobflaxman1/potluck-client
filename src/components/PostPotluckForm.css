@keyframes animation {
    from {left:  0px;}
    to {left: 200px;}
}


.field {
    border: none;
    padding: 0.5rem;
    width: 300px;
    margin: 5px;
    
    display: flex; 
    align-content: center;
    justify-content: center;
    flex-direction: column; 
}

.add-input {
    display: inline-block;
    border-radius: 10px;
    background-color: #EDF5E1;

    width: 50%;

}

.add-input:hover {
    background-color: #05386B;
}

.remove-input {

    border-radius: 10px;
    background-color: #EDF5E1;
    width: 25%;
}

.remove-input:hover {
    background-color: rgba(200, 45, 45, .6);
}

.submit-potluck-form {
    display: flex; 
    align-content: center;
    justify-content: space-between;
    flex-direction: column; 
    /* border: 5px solid #EDF5E1; */
    background-color: #8EE4AF;
    animation-name: animation;
    animation-duration: 2s;
    padding: 10px;
}

  
.no-suggestions {
    color: #999;
  }
  
.suggestions {
    border: 1px solid #999;
    list-style: none;
    max-height: 143px;
    overflow-y: auto;
    width: calc(300px + 1rem);
  }
  
  .suggestions li {
    padding: 0.5rem;
  }
  
.suggestion-active,
.suggestions li:hover {
    color: #fae042;
    cursor: pointer;
    font-weight: 700;
  }
  
.suggestions li:not(:last-of-type) {
    border-bottom: 1px solid #999;
  }

  .potluck-items {
      display: flex;
      justify-content: space-between;

  }
  