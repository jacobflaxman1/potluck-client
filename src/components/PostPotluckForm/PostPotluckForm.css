@keyframes animation {
    from {opacity: 0;}
    to {opacity: 100;}
}

label {
    margin-top: 5px;
    margin-bottom: 8px;
    color: #8F291F;
    font-weight: 550;
    font-size: 1.15rem;
    text-align: center;
}
.formheader {
    border-bottom: 2px solid #2471FF;
    margin-bottom: 15px;
}

.field {
    border-right: none;
    border-top: none;
    border-left: none;
    background-color:transparent;
    outline: none;
    color: #ffffff;
    font-size: large;
    font-weight: 550;
    caret-color: white;
    padding: 5px;
    margin-right: 20px;
}

.field:focus {
    border-bottom: 3px solid #2471FF;
}

.add-input {
	background: #ffffff;
	border: none;
	padding: 10px;
	border-bottom: 3px solid #5994FF;
	border-radius: 3px;
	color: #2471FF;
    font-size: 14px;
    align-self: center;
    margin-top: 15px;
}

.add-input:hover {
    background-color: #05386B;
}

.remove-input {
    font-weight: 700;
	color: #05386B;
	background: #ffffff;
	font-size: 14px;
    width: 8em;
    height: 2em;
    align-self: center;
    border-width: 1px 1px 3px;
    margin-bottom: 6px;
    border: none;

}

.remove-input:hover {
    background-color: rgba(200, 45, 45, .6);
    border: none;
}

.submit-potluck-form {
    display: flex; 
    align-content: center;
    justify-content: space-between;
    flex-direction: column; 
    width: 100%;
    padding: 2vh;
    border: none;
    cursor: pointer;
    border-radius: 4px 4px;
}

  
.no-suggestions {
    color: #999;
  }
  
.suggestions {
    /* border: 1px solid #05386B; */
    list-style: none;
    max-height: 143px;
    overflow-y: auto;
    width: calc(10rem);
  }
  
  .suggestions li {
    padding: 0.5rem;
  }
  
.suggestion-active,
.suggestions li:hover {
    color: #05386B;
    cursor: pointer;
    font-weight: 700;
  }
  
.suggestions li:not(:last-of-type) {
    border-bottom: 1px solid #999;
  }


.addpeople {
      display: flex;
      flex-direction: column;
  }

.additems {
      display: flex;
      flex-direction: column;
  }

.submit-post {
	background: none;
	border: none;
	padding: 10px;
	border: 3px solid #5994FF;
	border-radius: 3px;
	color:#05386B ;
    font-size: 14px;
    width: 5em;
    align-self: center;
    font-weight: 600;
}

.submit-post:hover {
    background: #05386B;
    color: #ffffff
}