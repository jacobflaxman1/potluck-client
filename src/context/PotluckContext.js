import React, { Component } from 'react'
import itemApiService from '../services/item-api-service'
import UserApiService from '../services/user-api-service'

const PotluckContext = React.createContext({
    potluck: [],
    items: [],
    error:null,
    authToken: null,
    users: [],
    adminUser: [],
    usersInPotluck: [],
    setEror: () => {},
    setPotluck: () => {},
    setItems: () => {},
    clearError: () => {},
    clearPotluck: () => {},
    saveAuthToken: () => {},
    clearAuthToken: () => {},
    addPostedPotluck: () => {},
    updateItem: () => {},
    getAllUserNames: () => {},
    getGuestUsersInPotluck: () => {},
    getAdminUserInPotluck: () => {},
    // LIST CONTEXT
    potluckList: [],
    user: '',
    setPotluckList: () => {},
    setUser: () => {},
    setShowForm: () => {},
    showForm: false,
    setExpanded: () => {},
    render: false,
    setRender: () => {},
    clearPotluckInList: () => {}
})

export default PotluckContext

export class PotluckProvider extends Component {
    state = {
        potluck: [],
        items: [],
        users: [],
        adminUser: [],
        usersInPotluck:[],
        error: null,
        authToken: null,
        showForm: false,
        //LIST STATE
        potluckList: [],
        user: '',
        render: false
    }

    setRender = () => {
        this.setState({ render: !this.state.render})
    }

    setShowForm = () => {
        this.setState({ showForm: !this.state.showForm })
    }


    setError = error => {
        console.error(error)
        this.setState({ error })
    }

    clearError = () => {
        this.setState({ error: null })
    }

    setPotluck = (potluck) => {
        this.setState({ potluck })
    }

    setItems = (items) => {
        this.setState({ items })
    }

    clearPotluck = () => {
        this.setPotluck([])
    }

    clearPotluckInList = (id) => {
        let newList = [...this.state.potluckList]
        let index = this.state.potluckList.findIndex(el => el.potluck_id === id)

        if(index !== -1) {
            newList.splice(index, 1)
            this.setState({potluckList: newList})
        }
        
    }

    saveAuthToken = (authToken) => {
        this.setState({ authToken })
    }

    clearAuthToken = () => {
        this.setState({ authToken: null })
    }

    addPostedPotluck = (newPotluck) => {
        this.setState({ potluckList: [...this.state.potluckList, {...newPotluck, expanded: true}] })
    }
    
    updateItem = (id) => {
        itemApiService.updateItem(id)
            .then((data) => {
                const newItemData = this.state.items.map((item) => {
                    if(id === item.item_id) {
                        return data
                    }
                    return item
                })
                this.setState({items: newItemData})
        })
    }

    getAllUserNames = () => {
        UserApiService.getUsers()
            .then((data) => {
                let userNames = data.rows.map(user => {
                    return user
                })
                this.setState({ users: userNames})
            })
    }
    getGuestUsersInPotluck = (id) => {
        console.log(id)
        UserApiService.getUsersByPotluck(id)
            .then((data) => {
                let userNames = data.guests.map(user => {
                    return user
                })
                this.setState({ usersInPotluck: userNames})
        })
    }
    getAdminUserInPotluck = (id) => {
        UserApiService.getUsersByPotluck(id) 
            .then((data) => {
                let adminUser = data.admin.map(ad => {
                    return ad
                })
                this.setState({ adminUser: adminUser})
            })
    }
    //LIST FUNCTIONS
    setPotluckList = potluckList => {
        const expandedPotluckList = potluckList.map(data => {
            return {...data, expanded: false}
        })

        this.setState({ potluckList: expandedPotluckList })
      }

    setUser = (user) => {
        this.setState({ user })
      }


    setExpanded = (id) => {
        const index = this.state.potluckList.findIndex(el => el.potluck_id === id)
        let newPotluckState = [...this.state.potluckList]

        newPotluckState[index] = {...newPotluckState[index], expanded: !newPotluckState[index].expanded}
        this.setState({ potluckList: newPotluckState })
    }

    render() {
        const value = {
            potluck: this.state.potluck,
            items: this.state.items,
            users: this.state.users,
            usersInPotluck: this.state.usersInPotluck,
            adminUser: this.state.adminUser,
            error: this.state.error,
            authToken: this.state.authToken,
            potluckList: this.state.potluckList,
            user: this.state.user,
            setItems: this.setItems,
            setError: this.setError,
            setPotluck: this.setPotluck,
            clearPotluck: this.clearPotluck,
            saveAuthToken: this.saveAuthToken,
            clearAuthToken: this.clearAuthToken,
            addPostedPotluck: this.addPostedPotluck,
            updateItem: this.updateItem,
            getAllUserNames: this.getAllUserNames,
            getGuestUsersInPotluck: this.getGuestUsersInPotluck,
            getAdminUserInPotluck: this.getAdminUserInPotluck,
            setPotluckList: this.setPotluckList,
            setUser: this.setUser,
            clearError: this.clearError,
            showForm: this.state.showForm,
            setShowForm: this.setShowForm,
            setExpanded: this.setExpanded,
            render: this.state.render,
            setRender: this.setRender,
            clearPotluckInList: this.clearPotluckInList

        }
        return(
            <PotluckContext.Provider value = {value}>
                {this.props.children}
            </PotluckContext.Provider>
        )
    }

}