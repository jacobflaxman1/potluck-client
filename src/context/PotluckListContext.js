import React, { Component } from 'react'

const PotluckListContext = React.createContext({
  potluckList: [],
  error: null,
  user: '',
  setError: () => {},
  clearError: () => {},
  setPotluckList: () => {},
  setUser: () => {}
})
export default PotluckListContext

export class PotluckListProvider extends Component {
  state = {
    potluckList: [],
    user: '',
    error: null,
  };

  setPotluckList = potluckList => {
    this.setState({ potluckList })
  }

  setError = error => {
    console.error(error)
    this.setState({ error })
  }

  clearError = () => {
    this.setState({ error: null })
  }

  setUser = (user) => {
    this.setState({ user })
  }

  render() {
    const value = {
      potluckList: this.state.potluckList,
      error: this.state.error,
      setError: this.setError,
      clearError: this.clearError,
      setPotluckList: this.setPotluckList,
      setUser: this.setUser
    }
    return (
      <PotluckListContext.Provider value={value}>
        {this.props.children}
      </PotluckListContext.Provider>
    )
  }
}
