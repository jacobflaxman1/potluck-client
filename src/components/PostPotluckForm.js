import React, { Component } from 'react'
import './PostPotluckForm.css'
import PotluckApiService from '../services/potluck-api-service'
import PotluckContext from '../context/PotluckContext'


export default class PostPotluckForm extends Component {
    static contextType = PotluckContext

    state = { error: null, values: [""], people: [""] }

    handleSubmit = e => {
        e.preventDefault()
        const { potluck_name } = e.target
        PotluckApiService.postPotluck({
            potluck_name: potluck_name.value,
            potluck_items: this.state.values,
            potluck_people: this.state.people
        })
        .then(potluck => {
            potluck_name.value = ''
            this.setState({ values: ['']})
            this.setState({ people: ['']})
            this.context.addPostedPotluck(potluck)
            this.props.history.push('/')
        })
        .catch(res => {
            this.setState({ error: res.error})
        })
    }
    //FOR VALUES IN STATE
    handleChange = (event, i) => {
        let values = [...this.state.values];
        let val = event.target.value;
        values[i] = val;
        this.setState({ values });
      };
      addClick = () => {
        this.setState(prevState => ({
          values: [...prevState.values, ""]
        }));
      };
      removeClick = i => {
        let values = [...this.state.values];
        values.splice(i, 1);
        this.setState({ values });
      };
      //FOR PEOPLE IN STATE
      handleChangePeople = (event, i) => {
        let people = [...this.state.people];
        let val = event.target.value;
        people[i] = val;
        this.setState({ people });
      };
      addClickPeople = () => {
        this.setState(prevState => ({
          people: [...prevState.people, ""]
        }));
      };
      removeClickPeople = i => {
        let people = [...this.state.people];
        people.splice(i, 1);
        this.setState({ people });
      };
      render() {
        const { error } = this.state;
        return (
          <form className="submit-potluck-form" onSubmit={this.handleSubmit}>
            <div> {error && <p>{error} </p>} </div>
            <label htmlFor="potluck-name"> Name Your Potluck:</label>
            <input name="potluck_name" required />
            <br />
            <label htmlFor="potluck-items"> Add Items to Your Potluck: </label>
            {this.state.values.map((el, i) => (
              <div key={i}>
                <input
                  type="text"
                  value={this.state.values[i]}
                  onChange={e => this.handleChange(e, i)}
                />{" "}
                &nbsp;&nbsp;
                <input
                  type="button"
                  value="remove"
                  onClick={() => this.removeClick(i)}
                />
              </div>
            ))}
            <input type="button" value="add more" onClick={() => this.addClick()} />
            <br />
            <label htmlFor="potluck-people"> Add People to Your Potluck: </label>
            {this.state.people.map((el, i) => (
              <div key={i}>
                <input
                  type="text"
                  value={this.state.people[i]}
                  onChange={e => this.handleChangePeople(e, i)}
                />{" "}
                &nbsp;&nbsp;
                <input
                  type="button"
                  value="remove"
                  onClick={() => this.removeClickPeople(i)}
                />
                <input type="button" value="add more" onClick={() => this.addClickPeople()} />
              </div>
            ))}
 
            <br />
            <button type="submit"> Post </button>
          </form>
        );
      }
    }