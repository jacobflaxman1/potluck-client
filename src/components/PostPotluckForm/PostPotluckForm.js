import React, { Component } from 'react'
import './PostPotluckForm.css'
import PotluckApiService from '../../services/potluck-api-service'
import PotluckContext from '../../context/PotluckContext'
import { ReactDOM } from 'react-dom'

export default class PostPotluckForm extends Component {
    static contextType = PotluckContext

    state = { 
        error: null,
        values: [""],
        people: [""],
        activeSuggestion: 0,
        filteredSuggestions: [],
        showSuggestions: false,
        indexOfInput: 0,
        focused: false,
        currentStep: 1
    }

    handleSubmit = e => {
        e.preventDefault()
        const { potluck_name } = e.target
        PotluckApiService.postPotluck({
            potluck_name: potluck_name.value,
            potluck_items: this.state.values,
            potluck_people: this.state.people
        })
        .then(potluck => {
            potluck_name.value = ''
            this.setState({ values: ['']})
            this.setState({ people: ['']})
            this.context.addPostedPotluck(potluck)
            this.context.setShowForm()
        })
        
        .catch(res => {
            this.setState({ error: res.error})
        })
    }
    //FOR VALUES IN STATE
    handleChange = (event, i) => {
        let values = [...this.state.values];
        let val = event.target.value;
        values[i] = val;
        this.setState({ values });
      };
      addClick = () => {
        this.setState(prevState => ({
          values: [...prevState.values, ""]
        }));
      };
      removeClick = i => {
        let values = [...this.state.values];
        values.splice(i, 1);
        this.setState({ values });
      };

      //SET STATE FOR PEOPLE IN STATE AND SUGGESTIONS
      handleChangePeople = (event, i) => {
        let people = [...this.state.people];
        let val = event.target.value;
        people[i] = val;
        this.setState({ people });
        //AUTOCOMPLETE BASED ON WHAT THEY ARE TYPING INTO INPUT
        const { users } = this.context

        let arrayOfUsers = users.map(user => {
            return Object.values(user)
        })
        const filteredSuggestions = arrayOfUsers.filter(
            suggestion => {
                return suggestion[0].toLowerCase().indexOf(val.toLowerCase()) > -1 }
        )
        this.setState({
            activeSuggestion: 0,
            filteredSuggestions,
            showSuggestions: true
        })
      };
      //ADD INPUT FIELD
      addClickPeople = () => {
        this.setState({people: [...this.state.people, ""]})
      };

      //DELETE INPUT FIELD
      removeClickPeople = i => {
        let people = [...this.state.people];
        people.splice(i, 1);
        this.setState({ people });
      };

      onClickSuggestion = (e, i) => {
        let { activeSuggestion, filteredSuggestions } = this.state
        let people = this.state.people
        activeSuggestion = e.target.getAttribute("data-index")
        people[this.state.indexOfInput] = (filteredSuggestions[activeSuggestion])
          this.setState({
              activeSuggestion: 0,
              filteredSuggestions: [],
              showSuggestions: false,
              people
          })
      }
      
      onFocus = (i) => {
        this.setState({ focused: true, indexOfInput: i })
      }
      onBlur = (i) => {
        this.setState({ focused: false, indexOfInput: i })
      }
      
      nextStep = () => {
        let currentStep = this.state.currentStep
        // If the current step is 1 or 2, then add one on "next" button click
        currentStep = currentStep >= 2? 3: currentStep + 1
        this.setState({
          currentStep: currentStep
        })
      }

      prevStep = () => {
        let currentStep = this.state.currentStep
        // If the current step is 2 or 3, then subtract one on "previous" button click
        currentStep = currentStep <= 1? 1: currentStep - 1
        this.setState({
          currentStep: currentStep
        })
      }

      render() {
        const { error, filteredSuggestions, activeSuggestion } = this.state;
        const { onClickSuggestion } = this
        let suggestionsListComponent;
        let noSuggestionsList;

        suggestionsListComponent = (
          <ul className="suggestions">
            {filteredSuggestions.map((suggestion, index) => {
              let className;
              if (index === activeSuggestion && this.state.focused && this.state.indexOfInput === index) {
                className = "suggestion-active";
              }
              return (
                <li data-index={index} className={className} key={suggestion} onClick={onClickSuggestion}>
                  {suggestion}
                </li>
              );
            })}
          </ul>
        );
        noSuggestionsList = (
          <ul className="suggestions">
          </ul>
        );
            console.log(this.state.values, this.state.people)
        return (
          <form className="submit-potluck-form" onSubmit={this.handleSubmit}>
            <div> {error && <p>{error} </p>} </div>
         {this.state.currentStep === 1 && <> <label htmlFor="potluck-name"> Name Your Potluck:</label>
            <input name="potluck_name" required className = 'field'/>
            <button onClick = {this.nextStep}> next step </button>
             </> }
            <br />
           {this.state.currentStep === 2 && <>
            <label htmlFor="potluck-items"> Add Items to Your Potluck </label>
            {this.state.values.map((el, i) => (
              <div key={i} className = 'additems'>
                <input className = 'field'  
                  type="text"
                  value={this.state.values[i]}
                  onChange={e => this.handleChange(e, i)}
                />{" "}
                &nbsp;&nbsp;
                <input className = 'remove-input'
                  type="button"
                  value="-"
                  onClick={() => this.removeClick(i)}
                />
              </div>
            ))}
            <input className = 'add-input' type="button" value="Add More Items" onClick={() => this.addClick()} />
            <button onClick = {this.nextStep}> next step </button>
            </>}
            <br />
           {this.state.currentStep === 3 && <>
            <label htmlFor="potluck-people"> Add Friends By UserName </label>
            {this.state.people.map((el, i) => (
                <div className = 'addpeople' key = {i}>
                    <input className = 'field' 
                        type="text"
                        value={this.state.people[i]}
                        onChange={e => this.handleChangePeople(e, i)}
                        onFocus = {() => this.onFocus(i)}
                        onBlur = {() => this.onBlur(i)}
                    />
                    {(this.state.indexOfInput === i) ? suggestionsListComponent : noSuggestionsList}
                    &nbsp;&nbsp;
                    <input className = 'remove-input'
                        type="button"
                        value="-"
                        onClick={() => this.removeClickPeople(i)}
                    />
                 </div>
            ))}
           <input className = 'add-input' type="button" value="Add Another Person" onClick={() => this.addClickPeople()} />
            </>}
            <br />
           <button type="submit" className = 'submit-post'> Create </button>
          </form>
        );
      }
    }
