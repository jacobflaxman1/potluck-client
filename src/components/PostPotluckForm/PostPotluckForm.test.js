import React from 'react';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { PotluckProvider } from '../../context/PotluckContext'
import PostPotluckForm from './PostPotluckForm';

describe(`Adds Potluck Form to dom without crashing`, () => {
    it('renders without crashing', () => {
        const wrapper = shallow(
        <PotluckProvider>
            <PostPotluckForm />
        </PotluckProvider>
    )
    .find('.submit-potluck-form')
    expect(toJson(wrapper)).toMatchSnapshot()
    })
})