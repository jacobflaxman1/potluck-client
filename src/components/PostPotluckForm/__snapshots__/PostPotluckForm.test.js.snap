// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adds Potluck Form to dom without crashing renders without crashing 1`] = `null`;
